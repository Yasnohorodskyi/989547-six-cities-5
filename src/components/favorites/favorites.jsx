import React, {Fragment} from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import {withDataLoading} from "../../hocks/with-data-loading";
import BookmarkCard from "../offer-card-bookmark/offer-card-bookmark";
import {getOffersByCity} from "../../store/reducers/app-data/app-data";
import propTypes from "prop-types";
import classNames from "classnames";

const Favorites = (props) => {
  const bookmarks = getOffersByCity(props.bookmarks);
  const userInfo = props.userInfo;
  const favorites = () => {
    const favoritesContainer = [];
    let initialIndex = 0;

    bookmarks.forEach((value, key) => {
      favoritesContainer.push({
        city: key,
        id: `${initialIndex}${key}`,
        offers: value
      });
      initialIndex++;
    });
    return favoritesContainer;
  };

  const renderContent = () => {
    if (bookmarks.size === 0) {
      return (
        <Fragment>
          <h1 className="visually-hidden">Favorites (empty)</h1>
          <div className="favorites__status-wrapper">
            <b className="favorites__status">Nothing yet saved.</b>
            <p className="favorites__status-description">Save properties to narrow down search or plan yor future trips.</p>
          </div>
        </Fragment>
      );
    }
    return (
      <Fragment>
        <h1 className="favorites__title">Saved listing</h1>
        <ul className="favorites__list">
          {favorites().map((offersByCity) => {
            return (
              <li key={offersByCity.id} className="favorites__locations-items">
                <div className="favorites__locations locations locations--current">
                  <div className="locations__item">
                    <a className="locations__item-link" href="#">
                      <span>{offersByCity.city}</span>
                    </a>
                  </div>
                </div>
                <div className="favorites__places">
                  {
                    offersByCity.offers.map((offer) => {
                      return <BookmarkCard key={offer.id} offer={offer}/>;
                    })
                  }
                </div>
              </li>
            );
          })}
        </ul>
      </Fragment>
    );
  };

  const mainSectionClass = bookmarks.size === 0
    ? `favorites--empty`
    : ``;

  return (
    <div className="page">
      <header className="header">
        <div className="container">
          <div className="header__wrapper">
            <div className="header__left">
              <Link className="header__logo-link" to="/">
                <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width={81} height={41} />
              </Link>
            </div>
            <nav className="header__nav">
              <ul className="header__nav-list">
                <li className="header__nav-item user">
                  <a className="header__nav-link header__nav-link--profile" href="#">
                    <div className="header__avatar-wrapper user__avatar-wrapper">
                    </div>
                    <span className="header__user-name user__name">{userInfo.email}</span>
                  </a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </header>
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className={classNames(`favorites`, mainSectionClass)}>
            {renderContent()}
          </section>
        </div>
      </main>
      <footer className="footer container">
        <a className="footer__logo-link" href="main.html">
          <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width={64} height={33} />
        </a>
      </footer>
    </div>
  );
};

const mapStateToProps = ({USER, DATA}) => ({
  isDataLoaded: USER.isUserStatusLoaded && DATA.areBookmarksLoaded,
  bookmarks: DATA.bookmarks,
  userInfo: USER.userInfo,
});

Favorites.propTypes = {
  bookmarks: propTypes.array.isRequired,
  userInfo: propTypes.object.isRequired
};

export {Favorites};
export default connect(mapStateToProps)(withDataLoading(Favorites));
