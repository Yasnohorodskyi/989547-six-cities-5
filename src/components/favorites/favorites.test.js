import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import configureStore from "redux-mock-store";

import {Favorites} from "./favorites";

const offers = [
  {
    id: 123,
    city: {
      name: `Amsterdam`
    },
    cost: 444,
    isFavorite: true,
    isPremium: false,
    previewImg: `https://picsum.photos/210`,
    rate: 4,
    title: `Bla bla bla`,
    type: `Room`
  },
  {
    id: 124,
    city: {
      name: `Amsterdam`
    },
    cost: 445,
    isFavorite: true,
    isPremium: false,
    previewImg: `https://picsum.photos/210`,
    rate: 4,
    title: `Hello, mello, shmello`,
    type: `House`
  },
];

const mockStore = configureStore()({
  USER: {
    isDataLoaded: false,
    userInfo: {
      email: `ibrohim@gmail.com`
    }
  },
  DATA: {
    bookmarks: offers
  }
});

const userInfo = {
  email: `ibrohim@gmail.com`
};

it(`Should favorites page render`, () => {
  const tree = renderer
  .create(
      <Provider store={mockStore}>
        <MemoryRouter>
          <Favorites
            bookmarks={offers}
            userInfo={userInfo}
          />
        </MemoryRouter>
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
