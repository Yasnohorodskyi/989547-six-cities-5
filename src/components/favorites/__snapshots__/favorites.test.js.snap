// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should favorites page render 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height={41}
              src="img/logo.svg"
              width={81}
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="#"
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  ibrohim@gmail.com
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--favorites"
  >
    <div
      className="page__favorites-container container"
    >
      <section
        className="favorites"
      >
        <h1
          className="favorites__title"
        >
          Saved listing
        </h1>
        <ul
          className="favorites__list"
        >
          <li
            className="favorites__locations-items"
          >
            <div
              className="favorites__locations locations locations--current"
            >
              <div
                className="locations__item"
              >
                <a
                  className="locations__item-link"
                  href="#"
                >
                  <span>
                    Amsterdam
                  </span>
                </a>
              </div>
            </div>
            <div
              className="favorites__places"
            >
              <article
                className="favorites__card place-card"
                onMouseOver={[Function]}
              >
                
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <a
                    href="/offer/123"
                    onClick={[Function]}
                  >
                    <img
                      className="place-card__image"
                      height="110"
                      src="https://picsum.photos/210"
                      width="150"
                    />
                  </a>
                </div>
                <div
                  className="place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        444
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button place-card__bookmark-button--active"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height={19}
                        width={18}
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      href="/offer/123"
                      onClick={[Function]}
                    >
                      Bla bla bla
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    Room
                  </p>
                </div>
              </article>
              <article
                className="favorites__card place-card"
                onMouseOver={[Function]}
              >
                
                <div
                  className="favorites__image-wrapper place-card__image-wrapper"
                >
                  <a
                    href="/offer/124"
                    onClick={[Function]}
                  >
                    <img
                      className="place-card__image"
                      height="110"
                      src="https://picsum.photos/210"
                      width="150"
                    />
                  </a>
                </div>
                <div
                  className="place-card__info"
                >
                  <div
                    className="place-card__price-wrapper"
                  >
                    <div
                      className="place-card__price"
                    >
                      <b
                        className="place-card__price-value"
                      >
                        €
                        445
                      </b>
                      <span
                        className="place-card__price-text"
                      >
                        / night
                      </span>
                    </div>
                    <button
                      className="place-card__bookmark-button button place-card__bookmark-button--active"
                      onClick={[Function]}
                      type="button"
                    >
                      <svg
                        className="place-card__bookmark-icon"
                        height={19}
                        width={18}
                      >
                        <use
                          xlinkHref="#icon-bookmark"
                        />
                      </svg>
                      <span
                        className="visually-hidden"
                      >
                        To bookmarks
                      </span>
                    </button>
                  </div>
                  <div
                    className="place-card__rating rating"
                  >
                    <div
                      className="place-card__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                      </span>
                    </div>
                  </div>
                  <h2
                    className="place-card__name"
                  >
                    <a
                      href="/offer/124"
                      onClick={[Function]}
                    >
                      Hello, mello, shmello
                    </a>
                  </h2>
                  <p
                    className="place-card__type"
                  >
                    House
                  </p>
                </div>
              </article>
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="main.html"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height={33}
        src="img/logo.svg"
        width={64}
      />
    </a>
  </footer>
</div>
`;
